### PR TITLE
speed up multi-val bin subset for bagging

### PR DESCRIPTION
**Pull request by guolinke**

### CODE REVIEW

**Comment by guolinke** [comment]
refer to #2783
BTW, I think these codes are very ugly, may need to refactor them.

**Comment by guolinke** [comment]
@StrikerRUS I cannot find the "rebuild" function in Travis anymore. 
Do you know where it is?

**Comment by StrikerRUS** [comment]
@guolinke 
>  I cannot find the "rebuild" function in Travis anymore.
Do you know where it is?

Try to re-login at Travis. Refer to https://github.com/microsoft/LightGBM/pull/2541#issuecomment-550504624.

File: src/treelearner/serial_tree_learner.h [hunk 1]
```diff
@@ -182,17 +200,12 @@ class SerialTreeLearner: public TreeLearner {
...
-  std::vector<char, Common::AlignmentAllocator<char, kAlignedSize>> is_data_in_leaf_;
   /*! \brief used to cache historical histogram to speed up*/
   HistogramPool histogram_pool_;
   /*! \brief config of tree learner*/
   const Config* config_;
-  std::vector<int> ordered_bin_indices_;
-  bool is_constant_hessian_;
-  std::unique_ptr<TrainingTempState> temp_state_;
-  bool is_hist_colwise_;
+  int num_threads_;
```
Comment (guolinke) [inline src/treelearner/serial_tree_learner.h]:
```suggestion
```

**Pull request merged**